syntax = "proto3";

package social_network;

// Post object
message Post {
  int32 id = 1;                
  string title = 2;             
  string description = 3;       
  int32 creator_id = 4;        
  string creation_date = 5;     
  string update_date = 6;       
  bool is_private = 7;          
  string tags = 8;     
}

// create a new post
message CreatePostRequest {
  string title = 1;
  string description = 2;
  int32 creator_id = 3;
  bool is_private = 4;
  string tags = 5;
}

// response for creating a post
message CreatePostResponse {
  int32 post_id = 1;  
  string message = 2;  
}

// update an existing post
message UpdatePostRequest {
  int32 post_id = 1;          
  string title = 2;           
  string description = 3;      
  bool is_private = 4;         
  string tags = 5;    
}

// response for updating a post
message UpdatePostResponse {
  int32 post_id = 1;   
  string message = 2;   
}

// delete a post
message DeletePostRequest {
  int32 post_id = 1;
}

// response for deleting a post
message DeletePostResponse {
  int32 post_id = 1;   
  string message = 2;   
}

// get post by id
message GetPostByIdRequest {
  int32 post_id = 1;
}

// get post info
message GetPostByIdResponse{
  Post post = 1;   
}

// just an empty for input
message Empty {}

// get list of posts
message GetPostsListResponse {
  repeated Post posts = 1;
}

service SocialNetworkService {
  rpc CreatePost (CreatePostRequest) returns (CreatePostResponse);
  rpc UpdatePost (UpdatePostRequest) returns (UpdatePostResponse);
  rpc DeletePost (DeletePostRequest) returns (DeletePostResponse);
  rpc GetPostById (GetPostByIdRequest) returns (GetPostByIdResponse);
  rpc GetPostsList (Empty) returns (GetPostsListResponse);
}
